### DISCUSSION TITLE
How to optimise Debug.assert

### DISCUSSION THREAD

**Discussion by mvaligursky** [playcanvas/engine#5473] [General]
currently we have this implementation

```
static assert(assertion, ...args) {
        if (!assertion) {
            console.error('ASSERT FAILED: ', ...args);
        }
    }
```

And typically it's called like this:

```
Debug.assert(condition, `Templated string with arguments ${argument}.`);
```

The problem here is that the templated string gets expanded before being passed to the assert function, which is expensive if this call is used in hot functions - we get lots of string concatenations.

Ideally, this would only take place when the assertion condition is false, which is when the string is logged out.

What are the options?

One I see is this:

```
// note that Debug.assert for this case is changed to simply unconditionally log args.
!assertion && Debug.assert(`Templated string with arguments ${argument}.`);
```

which works, but gives me this lint warning:

<img width="745" alt="Screenshot 2023-07-12 at 15 16 26" src="https://github.com/playcanvas/engine/assets/59932779/66ac8ecd-4a56-47eb-8fd0-ba12fca8ace8">

Is there something better?

**Comment by mvaligursky** [comment]
@willeastcott @slimbuck @kungfooman - thoughts?

**Reply by kungfooman** [reply]
```js
Debug.assert(condition, () => `Templated string with arguments ${argument}.`);
```

Maybe like this? If given a function, we can just call/resolve it. The extra syntax of `() =>` is a bit convoluted, but it could only be added in the hot path.

**Comment by mvaligursky** [comment]
@kungfooman - thanks! Yes, that was the first idea I looked at (and ChatGPT suggested), but it didn't feel clean enough, but perhaps it's fine. At the moment it feels a lot better than the suggestion I had in the initial post.

Any other ideas? If not, I'll go ahead with this I think.

**Reply by kungfooman** [reply]
We could also just remove the template string and make use of `...args`?

I guess another caveat with `!assertion && Debug.assert(...);` would be the AST shaking (remnants in the ES5 compilate).

**Comment by slimbuck** [comment]
Yep agree with @kungfooman 👍🏻

**Reply by kungfooman** [reply]
Maybe:

```js
Debug.assert(condition, 'Templated string with arguments', argument);
```

It's shorter, but it would still push more arguments on the stack every call... I didn't profile it, I just have a feeling `() =>` would be faster, since the arrow function isn't usually called anyway :thinking:

**Comment by GamesTocaBoca** [comment]
hey mvaligursky, in my opinin we can try "Instead of giving Debug the full message right away, we give Debug instructions on how to create the message, but only when it’s really needed.function check (condition, makeMessage) { if (!condition) {// If the condition is not true (false)
        console.error('Something went wrong: ', makeMessage()); https://gamestocaboca.com/
